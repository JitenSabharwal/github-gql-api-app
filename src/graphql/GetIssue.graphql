query GetIssue(
  $owner: String!
  $name: String!
  $issueNumber: Int!
  $afterComment: String
) {
  repository(owner: $owner, name: $name) {
    issue(number: $issueNumber) {
      ...Issue
    }
  }
}
