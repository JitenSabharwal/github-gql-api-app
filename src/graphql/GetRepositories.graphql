enum IssueState {
  Open
  Closed
}
query GetRepository(
  $owner: String!
  $name: String!
  $first: Int
  $last: Int
  $states: [IssueState!]
) {
  repository(owner: $owner, name: $name) {
    id
    nameWithOwner
    description
    url
    issues(states: $states, first: $first, last: $last) {
      edges {
        node {
          ...Issue
        }
      }
    }
  }
}

fragment Issue on Issue {
  title
  comments(first: $first) {
    totalCount
    edges {
      node {
        author {
          login
        }
        createdAt
      }
    }
  }
}
