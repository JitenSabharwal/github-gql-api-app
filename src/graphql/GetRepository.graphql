query GetRepository(
  $owner: String!
  $name: String!
  $first: Int
  $last: Int
  $after: String
  $before: String
  $states: [IssueState!]
) {
  repository(owner: $owner, name: $name) {
    id
    nameWithOwner
    description
    url
    issues(
      states: $states
      first: $first
      last: $last
      after: $after
      before: $before
    ) {
      totalCount
      pageInfo {
        hasNextPage
        startCursor
        endCursor
      }
      edges {
        node {
          ...IssueWithCommentsCount
        }
      }
    }
  }
}
