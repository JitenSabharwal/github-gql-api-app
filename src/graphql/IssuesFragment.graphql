fragment BasicIssueDetail on Issue {
  id
  title
  body
  state
  number
  createdAt
  updatedAt
  author {
    login
  }
}

fragment IssueWithCommentsCount on Issue {
  ...BasicIssueDetail
  comments(first: 10) {
    totalCount
  }
}

fragment Comment on Comment {
  bodyText
  author {
    login
  }
  createdAt
  updatedAt
}

fragment Issue on Issue {
  ...BasicIssueDetail
  comments(after: $afterComment, first: 50) {
    totalCount
    pageInfo {
      endCursor
      hasNextPage
      hasPreviousPage
    }
    edges {
      node {
        ...Comment
      }
    }
  }
}
