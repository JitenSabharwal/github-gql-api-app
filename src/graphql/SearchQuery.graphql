query SearchIssue($query: String!) {
  search(query: $query, type: ISSUE, first: 50) {
    issueCount
    edges {
      node {
        ...IssueWithCommentsCount
      }
      textMatches {
        property
      }
    }
  }
}
