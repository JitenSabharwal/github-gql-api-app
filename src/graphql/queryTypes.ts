import {
  QueryResult,
  ApolloError,
  LazyQueryResult,
  QueryHookOptions,
} from "@apollo/client"
import { useCallback, useMemo } from "react"

type Mapper<TData, TMappedData = TData> = (data: TData) => TMappedData

type SafeQueryResultLoadingStates<TData, TMappedData> =
  | { loading: true; error: undefined; data: undefined; mappedData: undefined }
  | { loading: false; error: undefined; data: TData; mappedData: TMappedData }
  | {
      loading: false
      error: ApolloError
      data: undefined
      mappedData: undefined
    }

type SafeQueryResult<TData = any, TVariables = any, TMappedData = TData> =
  SafeQueryResultLoadingStates<TData, TMappedData> & {
    isFetching: boolean
  } & QueryResult<TData, TVariables>

enum NetworkStatus {
  loading = 1,
  setVariables = 2,
  fetchMore = 3,
  refetch = 4,
  poll = 6,
  ready = 7,
  error = 8,
}
function isFetching(result: QueryResult | LazyQueryResult<any, any>) {
  return Boolean(
    result.networkStatus === NetworkStatus.refetch ||
      result.networkStatus === NetworkStatus.fetchMore
  )
}
function useSafeM<TData, TVariables, TMappedData = TData>(
  queryResult: QueryResult<TData, TVariables>,
  mapper: Mapper<TData | undefined, TMappedData | undefined>
) {
  const mappedData = useMemo(
    () => mapper(queryResult.data),
    [mapper, queryResult.data]
  )
  // TODO: shouldnt QR be stable object?
  return Object.assign(queryResult, {
    ...queryResult,
    mappedData,
    // for queries you want to participate, enable `notifyOnNetworkStatusChange`
    isFetching: isFetching(queryResult),
  }) as SafeQueryResult<TData, TVariables, TMappedData>
}

export function useSafeQuery<TData, TVariables, TMappedData = TData>(
  func: (
    baseOptions: QueryHookOptions<TData, TVariables>
  ) => QueryResult<TData, TVariables>,
  baseOptions: QueryHookOptions<TData, TVariables> & {
    memoMapper: Mapper<TData | undefined, TMappedData | undefined>
  }
) {
  return useSafeM(func(baseOptions), baseOptions.memoMapper)
}
