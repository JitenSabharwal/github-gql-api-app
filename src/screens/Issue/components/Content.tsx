import React from "react"
import styled from "styled-components"
import { IssueState } from "../../../generated/graphql"

const Root = styled.div``
const InfoWrapper = styled.div`
  margin-bottom: 24px;
`
const InfoItem = styled.div`
  margin-bottom: 8px;
`
const Label = styled.span`
  color: #777272;
  font-size: 13px;
  line-height: 24px;
`
const Description = styled.div`
  margin-bottom: 24px;
`

export interface ContentData {
  title: string
  body: string
  state: IssueState
  createdAt: Date
  updatedAt: Date
  author: string
}
type Props = {
  data: ContentData
}
export default function Content({ data }: Props) {
  return (
    <Root>
      <Label>Description: </Label>
      <Description>{data.body}</Description>
      <InfoWrapper>
        <InfoItem>
          <Label>State: </Label>
          <span>{data.state}</span>
        </InfoItem>
        <InfoItem>
          <Label>Updated at: </Label>
          <span>{data.updatedAt.toDateString()}</span>
        </InfoItem>
        <InfoItem>
          <Label>Created at: </Label>
          <span>{data.createdAt.toDateString()}</span>
        </InfoItem>
      </InfoWrapper>
    </Root>
  )
}
