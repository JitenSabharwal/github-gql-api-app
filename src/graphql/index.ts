import { useCallback } from "react"
import {
  useGetRepositoryQuery,
  GetRepositoryQueryVariables,
  GetRepositoryQuery,
  SearchIssueQueryVariables,
  useSearchIssueQuery,
  SearchIssueQuery,
  useGetIssueQuery,
  GetIssueQuery,
  IssueState,
  IssueCommentConnection,
} from "../generated/graphql"
import { useSafeQuery } from "./queryTypes"
import { Issue as IssueCardType } from "../components/molecules/IssueCard"
import { useParams } from "react-router-dom"

export const useGetRepository = ({
  owner,
  name,
  first,
  last,
  states,
  after,
  before,
}: GetRepositoryQueryVariables) => {
  const variables: GetRepositoryQueryVariables = {
    owner,
    name,
    states,
  }
  if (first) variables.first = first
  else if (last) variables.last = last
  if (after) variables.after = after
  if (before) variables.before = before
  return useSafeQuery(useGetRepositoryQuery, {
    variables,
    memoMapper: useCallback(
      (d: GetRepositoryQuery | undefined): MapGetRepo | undefined =>
        d?.repository ? mapInitalLoad(d) : undefined,
      []
    ),
  })
}

export type MapGetRepo = {
  repository: GetRepositoryQuery["repository"]
  issues: IssueCardType[]
  totalCount: number
}

function mapInitalLoad(d: GetRepositoryQuery): MapGetRepo {
  const issues = (d.repository?.issues.edges ?? []).map((n): IssueCardType => {
    const issue = n?.node
    return {
      id: issue?.number ?? 0,
      createdAt: new Date(),
      title: issue?.title ?? "",
      description: issue?.body ?? "",
      createdBy: issue?.author?.login ?? "",
      repository: d.repository?.nameWithOwner ?? "",
      commentCount: issue?.comments.totalCount ?? 0,
    }
  })
  return {
    repository: d.repository,
    totalCount: d.repository?.issues.totalCount ?? 0,
    issues,
  }
}

export function useSearchIssue({
  repo,
  state,
  text,
}: {
  repo: string
  state: IssueState
  text: string
}) {
  const variables: SearchIssueQueryVariables = {
    query: `repo:${repo} is:${state} in:title,body ${text}`,
  }
  return useSafeQuery(useSearchIssueQuery, {
    variables,
    memoMapper: useCallback(
      (d: SearchIssueQuery | undefined) =>
        d?.search.issueCount ? mapSearchResult(d, repo) : undefined,
      []
    ),
  })
}

function mapSearchResult(d: SearchIssueQuery, repo: string) {
  const issues_ = (d.search?.edges ?? []).map((x) => {
    if (x?.node?.__typename === "Issue")
      return {
        id: x.node?.number ?? 0,
        createdAt: new Date(),
        title: x.node?.title ?? "",
        description: x.node?.body ?? "",
        createdBy: x.node?.author?.login ?? "",
        repository: repo,
        commentCount: x.node?.comments.totalCount ?? 0,
      }
    return false
  })
  const issues = issues_.filter((i) => i) as IssueCardType[]
  return {
    issues,
    totalCount: issues?.length,
  }
}
interface MatchProps {
  id: string
  owner: string
  name: string
}

export const useGetIssue = () => {
  const { id, owner, name } = useParams<MatchProps>()
  return useSafeQuery(useGetIssueQuery, {
    variables: { issueNumber: +id, owner, name },
    memoMapper: useCallback(
      (d: GetIssueQuery | undefined) =>
        d?.repository?.issue ? mapIssue(d) : undefined,
      []
    ),
  })
}
type Comment = {
  body: string
  createdBy: string
  createdAt: Date
}
export type IssueContent = {
  nameWithOwner: string
  issue: {
    createdAt: string
    updatedAt: string
    title: string
    body: string
    author: string
    state: IssueState
    comments: Comment[]
  }
}
function mapIssue(d: GetIssueQuery): IssueContent {
  const { repository } = d
  return {
    nameWithOwner: repository?.nameWithOwner ?? "",
    issue: {
      createdAt: repository?.issue?.createdAt,
      updatedAt: repository?.issue?.updatedAt,
      title: repository?.issue?.title ?? "",
      body: repository?.issue?.body ?? "",
      author: repository?.issue?.author?.login ?? "",
      state: repository?.issue?.state ?? IssueState.Open,
      comments:
        repository?.issue?.comments.edges?.map((c) => {
          return {
            createdBy: c?.node?.author?.login ?? "",
            createdAt: new Date(c?.node?.updatedAt),
            body: c?.node?.bodyText ?? "",
          }
        }) ?? [],
    },
  }
}
