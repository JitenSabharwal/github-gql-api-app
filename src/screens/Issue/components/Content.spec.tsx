import Content, { ContentData } from "./Content"
import React from "react"
import { shallow } from "enzyme"
import { IssueState } from "../../../generated/graphql"

const mockData: ContentData = {
  createdAt: new Date(),
  updatedAt: new Date(),
  title: "Sample Issue",
  body: "This is just for testing purpose",
  author: "Jiten",
  state: IssueState.Open,
}
describe("Content", () => {
  let component: React.ReactElement
  beforeAll(() => (component = <Content data={mockData} />))

  it("Should render Content Component", () => {
    const container = shallow(component)
  })

  it("Should render the layout", () => {
    const container = shallow(component)
    expect(container.getElements()).toMatchSnapshot()
  })
})
