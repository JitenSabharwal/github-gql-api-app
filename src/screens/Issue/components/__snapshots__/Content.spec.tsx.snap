// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content Should render the layout 1`] = `
Array [
  <styled.div>
    <styled.div>
      This is just for testing purpose
    </styled.div>
    <styled.div>
      <styled.div>
        <styled.span>
          State: 
        </styled.span>
        <span>
          OPEN
        </span>
      </styled.div>
      <styled.div>
        <styled.span>
          Updated at: 
        </styled.span>
        <span>
          Wed May 26 2021
        </span>
      </styled.div>
      <styled.div>
        <styled.span>
          Created at: 
        </styled.span>
        <span>
          Wed May 26 2021
        </span>
      </styled.div>
    </styled.div>
  </styled.div>,
]
`;
